src/remote_gym/remote_environment_management.py

```diff
@@ -85,111 +95,239 @@ def start_as_remote_environment(
             )
 
+        # Pass action and obs layout
+        out_queue.put(action_spec)
+        out_queue.put(observation_spec)
+        initialized = True
+
+        while True:
+            action, reset = in_queue.get()
+            if reset is None:
+                break
+
+            if reset:
+                observation, info = env.reset()
+                reward = 0.0
+                terminated = False
+                truncated = False
+            else:
+                observation, reward, terminated, truncated, info = env.step(action)
+
+            rendering = env.render() if enable_rendering else None
+
+            out_queue.put((observation, reward, terminated, truncated, info, rendering))
+
+        env.close()
+    except Exception as e:
+        if not initialized:
+            out_queue.put(None)
+            out_queue.put(None)
+        out_queue.put(e)
+
+        logging.error("Stacktrace: %s", traceback.format_exc())
+
+
+class ProcessedEnv:
+    def __init__(self, args: dict, enable_rendering: bool, env_id: int):
+        self.in_queue = mp.Queue()
+        self.out_queue = mp.Queue()
+        self.should_reset = True
+
+        self.env_id = env_id
+        args["env_id"] = self.env_id
+
+        self.process = mp.Process(target=run_env_loop, args=(args, enable_rendering, self.in_queue, self.out_queue))
+        self.process.start()
+
+        self.action_spec, self.observation_spec = self.out_queue.get(), self.out_queue.get()
+
+        if self.action_spec is None or self.observation_spec is None:
+            raise self.out_queue.get()
+
+        self.action_manager = spec_manager.SpecManager(self.action_spec)
+        self.observation_manager = spec_manager.SpecManager(self.observation_spec)
+
+    def close(self):
+        self.in_queue.put((None, None))
+        self.process.join()
+
+    def step(self, action):
+        self.in_queue.put((action, self.should_reset))
+        self.should_reset = False
+        result = self.out_queue.get()
+        if isinstance(result, Exception):
+            raise result
+        return result
+
+    def reset(self):
+        self.should_reset = True
+
+
+class RemoteEnvironmentService(dm_env_rpc_pb2_grpc.EnvironmentServicer):
+    """Runs the environment as a gRPC EnvironmentServicer."""
+
+    def __init__(self, default_args: dict, enable_rendering: bool):
+        self.default_args = default_args
+        self.enable_rendering = enable_rendering
+        self.environments = {}
+
+        self.env_ids = [i for i in range(1024)][::-1]  # TODO: Magic number
+
+    def get_environment(self, user: str) -> ProcessedEnv:
+        if user not in self.environments:
+            raise ValueError(f"Environment for user {user} does not exist.")
+        return self.environments[user]
+
+    def new_environment(self, user: str, args: dict):
+        # We do not permit custom repositories for security reasons
+        # TODO: evaluate risk of custom entrypoints
```

Comment:
I think custom entrypoints do not pose an additional security issue *if you do not permit usage of custom repositories*.
In this case changing the entrypoint is the same as changing the code of the entrypoint file.

Unless it is possible to point to files outside the repository, because in that case it indeed is a security risk 😅

Reply:
Just in case I will enforce relative paths.

Reply:
Where are you enforcing relative paths? Haven't found it yet.

Reply:
Nevermind, found it.